### PR TITLE
Command:

### PR DESCRIPTION
**Pull request by TurretBox**
- added Custom Command Name
CommandDiscovery:
  - added Function to get All Assemblies in the Unity Project, and added that in the Build()
  - Tried to add a Execute function for Non-Static methodes, but we have to get the Type, so we can make an Activator/Instance of that class with the Methode
ConsoleWindow:
  - added Mini ToolBar
  - added Copy/Paste
  - Fixed so the GUISkin("named Layout") is being globaly used (but this makes the scrollBars black, because the GUISkin has that as the image) SEE LINE: 137
  - added Preferences for the ConsoleWindow, this only contains a toggle for AutoScroll
  - added Auto Scroll to the bottom when you are typing/ when you are executing a new line
  - added Log from default Unity Console

### CODE REVIEW

File: Assets/UPM/Console/Command.cs [hunk 1]
```diff
@@ -4,9 +4,13 @@
...
 using UnityEngine;
 
-
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = true, Inherited = false)]
 public class CommandAttribute : System.Attribute
 {
+    public string CustomName = ""; // maybe we wanna make a command named something that already exists, we can just add a custom name and use that instead :)
+                              // like this [Command(CustomName = "NameMe")]
+
+    public bool OnlyIngame = false; // maybe we make it so commands marked with this can only be run while the game is playing
```
Comment (runewake2) [inline Assets/UPM/Console/Command.cs:13]: Not sure about this. The Editor folder can be used to filter out commands from the published game, but can you give an example of an In Game Only command? What value is added by adding this restriction?

File: Assets/UPM/Console/CommandDiscovery.cs [hunk 2]
```diff
@@ -23,7 +27,7 @@ public static CommandDiscovery Build(params Assembly[] assemblies)
                 {
                     foreach (var attribute in method.GetCustomAttributes(false).OfType<CommandAttribute>())
                     {
-                        var command = new Command(method, method.Name);
+                        var command = new Command(method, attribute.CustomName == "" ? method.Name : attribute.CustomName); // added so we can have a custom Name :)
```
Comment (runewake2) [inline Assets/UPM/Console/CommandDiscovery.cs]: `String.IsNullOrEmpty(attribute.CustomName)` might be a bit more clear and cover the null case.

File: Assets/UPM/Console/CommandDiscovery.cs [hunk 3]
```diff
@@ -55,15 +68,30 @@ public void Invoke(string methodName)
...
     {
-        program.Invoke(null, arguments.ToArray());
+        if (program.IsStatic)
+        {
+            program.Invoke(null, arguments.ToArray());
+        }
+        else
+        {
+            //TODO fix non Static, because then we need to execute with the class that has this Methode/Program
+            object instance = Activator.CreateInstance(activator); // activator = Type = the class/Type that has the Methode ?
```
Comment (runewake2) [inline Assets/UPM/Console/CommandDiscovery.cs:91]: I'm not sure I understand the use case for this. Can give an example?

File: Assets/UPM/Editor/Console/ConsoleWindow.cs [hunk 4]
```diff
@@ -8,66 +8,224 @@
 [ExecuteInEditMode]
 public class ConsoleWindow : EditorWindow
 {
-    private StringBuilder currentLine = new StringBuilder();
-    private List<string> previousCommands = new List<string>();
-    private Vector2 scrollPosition = Vector2.zero;
+    /// <summary>
+    /// Clipboard: makes it so we can Copy and Paste stuff :)
+    /// </summary>
+    public static string clipboard
```
Comment (runewake2) [inline Assets/UPM/Editor/Console/ConsoleWindow.cs:14]: Nice use of a property to wrap the clipboard functionality!

File: Assets/UPM/Editor/Console/ConsoleWindow.cs [hunk 5]
```diff
@@ -8,66 +8,224 @@
...
+    private bool ConsoleLogEnabled = true; // should this console display stuff from the default unity console =?
+    private static bool ScrollToBottomOnNewCommand = true; // should the scrollview scroll to the bottom when entering a new command ?
+    private Vector2 scrollPosition = Vector2.zero; // current ScrollPosition
     public GUISkin layout;
 
     [MenuItem("Tools/Unity Shell %`")]
     static void InitializeWindow()
     {
         var window = GetWindow<ConsoleWindow>("Unity Shell", true);
+        window.titleContent.tooltip = "Unity Shell"; // fix me
```
Comment (runewake2) [inline Assets/UPM/Editor/Console/ConsoleWindow.cs]: What's broken?

File: Assets/UPM/Editor/Console/ConsoleWindow.cs [hunk 6]
```diff
@@ -8,66 +8,224 @@
...
+    private static bool PreferencesLoaded = false;
+
+    [PreferenceItem("Unity Shell")]
+    public static void PreferencesGUI()
+    {
+        //get values    (happens on start up)
+        if (!PreferencesLoaded)
+        {
+            //editorprefs something
+            EditorPrefs.GetBool("ScrollToBottmWhenNewCommand", true);
```
Comment (runewake2) [inline Assets/UPM/Editor/Console/ConsoleWindow.cs:50]: ScrollToBott**o**mWhenNewCommand

**Review by runewake2 — COMMENTED** [review]
A few quick questions and some minor changes for edge cases.

File: Assets/UPM/Console/CommandDiscovery.cs (hunk 2 above)
Comment (TurretBox) [reply Assets/UPM/Console/CommandDiscovery.cs]: yepp

File: Assets/UPM/Console/Command.cs (hunk 1 above)
Comment (TurretBox) [reply Assets/UPM/Console/Command.cs:13]: example: using the console ingame. x: a multiplayer game, spawning in items or changing something on a server?

we can add it later i guess

File: Assets/UPM/Editor/Console/ConsoleWindow.cs (hunk 5 above)
Comment (TurretBox) [reply Assets/UPM/Editor/Console/ConsoleWindow.cs]: maybe it should be "Unity Shell: In editor and InGame console"

File: Assets/UPM/Editor/Console/ConsoleWindow.cs (hunk 6 above)
Comment (TurretBox) [reply Assets/UPM/Editor/Console/ConsoleWindow.cs:50]: yes :)

File: Assets/UPM/Editor/Console/ConsoleWindow.cs (hunk 5 above)
Comment (TurretBox) [reply Assets/UPM/Editor/Console/ConsoleWindow.cs]: nothing is broken

File: Assets/UPM/Editor/Console/ConsoleWindow.cs (hunk 4 above)
Comment (TurretBox) [reply Assets/UPM/Editor/Console/ConsoleWindow.cs:14]: tnx :)

File: Assets/UPM/Console/CommandDiscovery.cs (hunk 3 above)
Comment (TurretBox) [reply Assets/UPM/Console/CommandDiscovery.cs:91]: this is the thing that I didnt fix last time I made a Console/Terminal back in Unity 5.3

the activator should be the Type of the class that has the methode (not sure)

File: Assets/UPM/Console/Command.cs (hunk 1 above)
Comment (runewake2) [reply Assets/UPM/Console/Command.cs:13]: I was thinking of this as commands for "Everywhere" and commands for "The Editor" I think your addition of a third "In Game" makes sense. Sort of a Venn Diagram of command capabilities. We'll have to figure out what the default is later. I will probably change this to an Enum in a quick patch after the PR.

Comment (TurretBox) [reply Assets/UPM/Console/Command.cs:13]: go ahead :)

File: Assets/UPM/Console/CommandDiscovery.cs (hunk 3 above)
Comment (runewake2) [reply Assets/UPM/Console/CommandDiscovery.cs:91]: But what scenario exists where you'd want a non-static command. I can't think of any off the top of my head. I can't think of a scenario where you'd write `SomeGameObject.Translate-Position (0,0,0)`. I picture that as `Find 'Some Game Object' > Translate-Position (0,0,0)` Is there something I'm not considering?

**Pull request merged**